### PR TITLE
Create Object based generalization for reasoning (Phyre).md

### PR DESCRIPTION
**Pull request by sarikasu**

### CODE REVIEW

File: tasks/Object based generalization for reasoning (Phyre).md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+Deep Q-network (DQN). As before, the DQN agent collects a set of observation-action-reward triplets by randomly sampling actions and running them through the simulator. The agent trains a deep network on the resulting data to predict the reward for an observation-action pair. Following [4], we train the network by minimizing the cross-entropy between the soft prediction and the observed reward. During training, we sample batches with an equal number of positive and negative triplets.
+
```
Comment (joycewang3) [inline tasks/Object based generalization for reasoning (Phyre).md:26]: Should we also include the fifth baseline agent mentioned:

Deep Q-network with online learning (DQN-O). Akin to MEM-O, this agent uses rewards from test tasks to perform online updates. After finishing a test task, the agent performs a number of gradient descent updates using examples obtained from that task. The updated model is then used for the next test task. The number of updates and corresponding learning rate are set via cross-validation.

File: tasks/Object based generalization for reasoning (Phyre).md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+Non-parametric agent (MEM). At training time, this agent generates a set of R random actions and uses the simulator to check if each of these actions can solve each of the training tasks. For each action a, the agent computes pa: the fraction of training tasks that the action solves. The agent then sorts the R actions by pa (highest to lowest), and tries them in this order at test time. This agent is non-parametric because it uses a list of “memorized” actions at test time.
+
+In the cross-template setting, the test tasks come from previously unseen task templates and this simple agent cannot relate them to tasks seen during training. It therefore uses the same action ranking for all tasks and ignores the observation of the initial state. In the within-template setting, each test task comes from a task template that was seen during training. In this case, we give the agent access to the task template id for each test task. The agent maintains a per-task-template ranking of the R actions. The same set of actions is shared across all templates; only the ranking changes. The set of actions attempted on each task may vary because invalid actions are ignored; see Section 3.2.
```
Comment (joycewang3) [inline tasks/Object based generalization for reasoning (Phyre).md:21]: I think I will remove the "see Section 3.2" reference

File: tasks/Object based generalization for reasoning (Phyre).md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+Non-parametric agent with online learning (MEM-O). This agent has the same training phase as the non-parametric agent, but continues to learn online at test time. Specifically, after finishing each test task (either successfully or unsuccessfully), the agent updates pa based on the reward received for each action a in the subset of the actions it attempted. The updated ranking is used when the next task is attempted. Such online updates are beneficial, in particular, in the cross-template setting because they allow the agent to learn something about the tasks in the previously unseen templates. We use cross-validation to tune the relative weight of the update on each train-val fold (see Section 4.2).
```
Comment (joycewang3) [inline tasks/Object based generalization for reasoning (Phyre).md:23]: Same here for the reference see Section 4.2

File: tasks/Object based generalization for reasoning (Phyre).md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Deep Q-network (DQN). As before, the DQN agent collects a set of observation-action-reward triplets by randomly sampling actions and running them through the simulator. The agent trains a deep network on the resulting data to predict the reward for an observation-action pair. Following [4], we train the network by minimizing the cross-entropy between the soft prediction and the observed reward. During training, we sample batches with an equal number of positive and negative triplets.
+
+# Procedure
+## Steps
```
Comment (joycewang3) [inline tasks/Object based generalization for reasoning (Phyre).md:28]: I think I would also add in the description of the two settings in the steps:

Task templates are used to measure an agent’s generalization ability in two settings. In the within-template setting, an agent trains on a subset of tasks in the template and is evaluated on the remaining tasks within that template. To measure cross-template generalization (in the cross-template generalization setting), test tasks are selected exclusively from templates that were not used for training.

File: tasks/Object based generalization for reasoning (Phyre).md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+Deep Q-network (DQN). As before, the DQN agent collects a set of observation-action-reward triplets by randomly sampling actions and running them through the simulator. The agent trains a deep network on the resulting data to predict the reward for an observation-action pair. Following [4], we train the network by minimizing the cross-entropy between the soft prediction and the observed reward. During training, we sample batches with an equal number of positive and negative triplets.
```
Comment (joycewang3) [inline tasks/Object based generalization for reasoning (Phyre).md:25]: Same here for "Following [4]"

**Review by sarikasu — COMMENTED** [review]
Updated based on feedback from Joyce

**Review by joycewang3 — APPROVED** [review]
